distribution/binary/compiler.bzl

```diff
@@ -0,0 +1,116 @@
+# Compile targets using a bazel-defined compilation mode.
+#
+# This is useful for building packages comprised of binaries built in
+# multiple modes - eg a release containing both `opt` and `dbg` builds.
+#
+# In this case the cli-supplied `-c ...` flag is ignored.
+#
+# NB: Avoid use of this as it creates a large sub-graph independent
+#  of the normal bazel build graph.
+#
+#  See https://bazel.build/extending/config#memory-performance-considerations
+#
+# The `bundled` rule allows you to define a set of compilation targets that you wish to
+# be compiled with compilation settings defined in bazel, that would normally be
+# set on the command line.
+#
+# The `mode` sets the `compilation_mode` - one of `dbg`, `opt`, or `fastbuild`.
+#
+# You can define the `targets` to be compiled, along with the destination `path`
+# of the compiled binary.
+#
+# You can also define lists of `stripopts` and `linkopts`.
+#
+# For example, you can bundle some stripped binaries together in `opt` mode:
+#
+# ```starlark
+#
+# bundled(
+#     name = "envoy",
+#     mode = "opt",
+#     stripopts = ["--strip-all"],
+#     targets = {
+#         "//distribution:envoy-binary": "envoy",
+#         "//distribution:envoy-contrib-binary": "envoy-contrib",
+#         "//external:su-exec": "utils/su-exec",
+#     },
+# )
+#
+#
+# The bundled output can then be packaged together with other, differently compiled, output:
+#
+# ```starlark
+#
+# bundled(
+#     name = "envoy-debug",
+#     mode = "dbg",
+#     targets = {
+#         "//distribution:envoy-binary": "envoy",
+#         "//distribution:envoy-dwarf": "envoy.dwp",
+#         "//distribution:envoy-contrib-binary": "envoy-contrib",
+#         "//distribution:envoy-contrib-dwarf": "envoy-contrib.dwp",
+#     },
+# )
+#
+# ```
+#
+
+def _compilation_config_impl(settings, attr):
+    _ignore = settings
+    return {
+        "//command_line_option:compilation_mode": attr.mode,
+        "//command_line_option:linkopt": attr.linkopts,
+        "//command_line_option:stripopt": attr.stripopts,
+    }
+
+compilation_config = transition(
+    implementation = _compilation_config_impl,
+    inputs = [],
+    outputs = [
+        "//command_line_option:compilation_mode",
+        "//command_line_option:linkopt",
+        "//command_line_option:stripopt",
```

Comment:
> should these be additive instead?

preferably not - because we want to control the settings for anything built below here

> maybe we already get some of that naturally since many linkopts are set directly on targets

yep there was a bug around merging these - i believe that works now

>  but if you build one of these target and pass --linkopt=-v is that also ignored?

hopefully, but possibly not - the goal here is to build package sets with their own flags and i guess to make it more deterministic - i tested this quite a bit and this seems to do what is required - but this is an area with a lot of corners, so i mostly just tested for things i did/not want

Reply:
But doesn't that mean we would also miss some of our own required ones like this? https://github.com/envoyproxy/envoy/blob/17844b10a3e57e411f3b11c595445327d6ac4d49/.bazelrc#L72

Reply:
hmm, i think you are right - i think there is a lot of things that would not work if it it wasnt additive

what is defo not is `COMPILATION_MODE` - i have seen tickets/hacks to keep the global - ie `-c opt/dbg/etc` if its set rather than overwriting it

Reply:
this actually makes me realize there is also a bit of a weird case, if you pass `-c dbg` and we `select()` on that, will it get the correct value in the select or not? maybe it's smart enough to since the `config_setting` should get the transition here as well

Reply:
> will it get the correct value in the select or not? 

well i guess it depends on expectation

i think this will works as ~expected - it will get the transitioned value not the global anywhere in the graph below here - that is the one thing i tested to death

Reply:
>  maybe it's smart enough to since the config_setting should get the transition here as well

yeah - i think we have the same expectation - this is what i tested and works i reckon

Reply:
this does work as expected right? https://github.com/envoyproxy/envoy/pull/26458/files#r1153601790

Reply:
can i check with `ldd` on the binary?

Reply:
(i believe so on the basis that when i tried to incorrectly set this it dumped the linker args and iirc it was the usual envoy ones)